src/group_theory/exponent.lean

```diff
@@ -0,0 +1,167 @@
+* `add_monoid.exponent_exists` the additive version of `monoid.exponent_exists`.
+* `add_monoid.exponent` the additive version of `monoid.exponent`.
+
+## Main results
+
+* `lcm_order_eq_exponent`: For a finite group `G`, the exponent is equal to the `lcm` of the order
+  of its elements.
+
+## TODO
+* Compute the exponent of cyclic groups.
+* Refactor the characteristic of a ring to be the exponent of its underlying additive group.
+-/
+
+universe u
+
+variables (G : Type u) [monoid G]
+
+open_locale classical
+
+namespace monoid
+
+/-A predicate on a monoid saying that there is a positive integer `n` such that `g ^ n = 1`
+  for all `g`.-/
+@[to_additive]
+def exponent_exists  := ∃ n, 0 < n ∧ ∀ g : G, g ^ n = 1
+
+/-The exponent of a group is the smallest positive integer `n` such that `g ^ n = 1` for all `g ∈ G`
+if it exists, otherwise it is zero by convention.-/
+@[to_additive]
+noncomputable def exponent :=
+if h : exponent_exists G then nat.find h else 0
+
+@[to_additive exponent_nsmul_eq_zero]
+lemma pow_exponent_eq_one (g : G) : g ^ exponent G = 1 :=
+begin
+  by_cases exponent_exists G,
+  { simp_rw [exponent, dif_pos h],
+    exact (nat.find_spec h).2 g},
+  { simp_rw [exponent, dif_neg h, pow_zero] }
+end
+
+@[to_additive nsmul_eq_mod_exponent]
+lemma pow_eq_mod_exponent {n : ℕ} (g : G): g ^ n = g ^ (n % exponent G) :=
+calc g ^ n = g ^ (n % exponent G + exponent G * (n / exponent G)) : by rw [nat.mod_add_div]
+  ... = g ^ (n % exponent G) : by simp [pow_add, pow_mul, pow_exponent_eq_one]
+
+@[to_additive]
+lemma exponent_pos_of_exists (n : ℕ) (hpos : 0 < n)
+(hG : ∀ g : G, g ^ n = 1) : 0 < exponent G :=
+begin
+  have h : ∃ n, 0 < n ∧ ∀ g : G, g ^ n = 1,
+      { exact ⟨n, hpos, hG⟩ },
+  rw [exponent, dif_pos],
+  exact (nat.find_spec h).1,
+end
+
+@[to_additive]
+lemma exponent_min' (n : ℕ) (hpos : 0 < n) (hG : ∀ g : G, g ^ n = 1) :
+  exponent G ≤ n :=
+begin
+  rw [exponent, dif_pos],
+  { apply nat.find_min',
+    exact ⟨hpos, hG⟩ },
+  { exact ⟨n, hpos, hG⟩ },
+end
+
+@[to_additive]
+lemma exponent_min (m : ℕ) (hpos : 0 < m) (hm : m < exponent G) : ∃ g : G, g ^ m ≠ 1 :=
+begin
+  by_contradiction,
+  push_neg at h,
+  have hcon : exponent G ≤ m := exponent_min' G m hpos h,
+  linarith,
+end
+
+lemma exp_punit_eq_one : exponent punit = 1 :=
+begin
+  apply le_antisymm,
+  { apply exponent_min' _ _ nat.one_pos,
+    intro g,
+    refl },
+  { apply nat.succ_le_of_lt,
+    apply exponent_pos_of_exists _ 1 (nat.one_pos),
+    intro g,
+    refl },
+end
+
+@[to_additive add_order_dvd_exponent]
+lemma order_dvd_exponent (g : G) : (order_of g) ∣ exponent G :=
+begin
+  exact order_of_dvd_of_pow_eq_one (pow_exponent_eq_one G g)
+end
+
+@[to_additive]
+lemma exponent_dvd_of_forall_pow_eq_one (n : ℕ) (hpos : 0 < n) (hG : ∀ g : G, g ^ n = 1) :
+  exponent G ∣ n :=
+begin
+  apply nat.dvd_of_mod_eq_zero,
+  by_contradiction h,
+  have h₁ := nat.pos_of_ne_zero h,
```

Comment:
this is just a cool trick instead of a suggestion but you could also write this line as `have h₁ : 0 < _ := h.bot_lt` with some cool dot notation tricks :)

Reply:
That doesn't seem to work, but it might just be a small typo?

Reply:
ahh, it doesn't work because it's `¬ _=_`, instead of `_≠_`!